lib/pry-doc/pry_ext/show_source_with_c_internals/code_fetcher.rb

```diff
@@ -0,0 +1,125 @@
+require 'fileutils'
+require_relative 'c_file'
+require_relative 'symbol_extractor'
+
+module Pry::CInternals
+  class CodeFetcher
+    include Pry::Helpers::Text
+
+    class << self
+      attr_accessor :ruby_source_folder
+    end
+
+    # normalized
+    def self.ruby_version() RUBY_VERSION.tr(".", "_") end
+
+    self.ruby_source_folder = File.join(File.expand_path("~/.pry.d"), "ruby-#{ruby_version}")
+
+    attr_reader :line_number_style
+    attr_reader :symbol_extractor
+
+    def initialize(line_number_style: nil)
+      @line_number_style = line_number_style
+      @symbol_extractor = SymbolExtractor.new(self.class.ruby_source_folder)
+    end
+
+    def fetch_all_definitions(symbol)
+      infos = self.class.symbol_map[symbol]
```

Comment:
I'm struggling to understand what `infos` holds (and I can't find relevant tests). Is there a better name for this variable?

Reply:
@kyrylo `info` holds the information for a given symbol, so `symbol_map["foo"]` will return the information about the `foo` symbol, i.e its source_location and the type of symbol it is, i.e funcitno or struct or enum

in the case of `infos` it's because a symbol can be defined multiple times, we can have multiple functions called `foo` (if they're all static, for example) and the multiple definitions can be returned using a `show-source -a foo`

otherwise it just defaults to showing `infos.first`, i.e the first definition of that symbol

Reply:
Looks like with this call (` self.class.symbol_map`) you invoke `parse_tagfile` every single time, which is probably not good from the performance point of view.

Reply:
> @kyrylo info holds the information for a given symbol, so symbol_map["foo"] will return the information about the foo symbol, i.e its source_location and the type of symbol it is, i.e funcitno or struct or enum

Yeah, makes sense. However, the name `infos` is still super generic. `source_locations` is long, but it seems to be more descriptive (at least to me).

Reply:
@kyrylo but `parse_tagfile` uses `||=` so it doesn't acdtually execute each time